File: examples/protocol-bridges/spiffe/aport_client.go
package spiffebridge

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// VerificationRequest is the payload sent to APort for policy verification.
type VerificationRequest struct {
	PolicyID string         `json:"policy_id"`
	AgentID  string         `json:"agent_id"`
	Context  map[string]any `json:"context,omitempty"`
}

// VerificationResult is the normalized response returned by APort.
type VerificationResult struct {
	Verified bool           `json:"verified"`
	PolicyID string         `json:"policy_id"`
	AgentID  string         `json:"agent_id"`
	Passport map[string]any `json:"passport,omitempty"`
	Message  string         `json:"message,omitempty"`
	Details  map[string]any `json:"details,omitempty"`
	Raw      map[string]any `json:"raw,omitempty"`
}

// APortVerifier abstracts APort verification for tests and alternate SDKs.
type APortVerifier interface {
	Verify(context.Context, VerificationRequest) (VerificationResult, error)
}

// HTTPAPortClient is a minimal APort verification client for Go integrations.
type HTTPAPortClient struct {
	baseURL    string
	apiKey     string
	httpClient *http.Client
	userAgent  string
}

// NewHTTPAPortClient returns a client that calls APort's policy verification endpoint.
func NewHTTPAPortClient(baseURL, apiKey string, timeout time.Duration) (*HTTPAPortClient, error) {
	parsed, err := url.Parse(strings.TrimRight(baseURL, "/"))
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return nil, fmt.Errorf("invalid APort base URL: %q", baseURL)
	}

	if timeout <= 0 {
		timeout = defaultTimeout
	}

	return &HTTPAPortClient{
		baseURL:    strings.TrimRight(baseURL, "/"),
		apiKey:     apiKey,
		httpClient: &http.Client{Timeout: timeout},
		userAgent:  "aport-spiffe-bridge/1.0",
	}, nil
}

// Verify verifies an APort agent against a policy pack.
func (c *HTTPAPortClient) Verify(ctx context.Context, req VerificationRequest) (VerificationResult, error) {
	if strings.TrimSpace(req.PolicyID) == "" {
		return VerificationResult{}, fmt.Errorf("APort policy ID is required")
	}
	if strings.TrimSpace(req.AgentID) == "" {
		return VerificationResult{}, fmt.Errorf("APort agent ID is required")
	}

	body, err := json.Marshal(req)
	if err != nil {
		return VerificationResult{}, fmt.Errorf("marshal verification request: %w", err)
	}

	endpoint := fmt.Sprintf("%s/api/verify/policy/%s", c.baseURL, url.PathEscape(req.PolicyID))
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(body))
	if err != nil {
		return VerificationResult{}, fmt.Errorf("build verification request: %w", err)
	}

	httpReq.Header.Set("Accept", "application/json")
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("User-Agent", c.userAgent)
	if c.apiKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+c.apiKey)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return VerificationResult{}, fmt.Errorf("call APort verification endpoint: %w", err)
	}
	defer resp.Body.Close()

	var payload map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return VerificationResult{}, fmt.Errorf("decode APort verification response: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return VerificationResult{}, fmt.Errorf("APort verification failed with HTTP %d: %s", resp.StatusCode, messageFrom(payload))
	}

	return normalizeVerificationResult(payload, req), nil
}

func normalizeVerificationResult(payload map[string]any, req VerificationRequest) VerificationResult {
	verified := boolFrom(payload["verified"]) || boolFrom(payload["success"])

	return VerificationResult{
		Verified: verified,
		PolicyID: stringFromAny(payload["policy"], stringFromAny(payload["policy_id"], req.PolicyID)),
		AgentID:  stringFromAny(payload["agent_id"], req.AgentID),
		Passport: mapFromAny(payload["passport"]),
		Message:  stringFromAny(payload["message"], ""),
		Details:  mapFromAny(payload["details"]),
		Raw:      payload,
	}
}

func messageFrom(payload map[string]any) string {
	if message := stringFromAny(payload["message"], ""); message != "" {
		return message
	}
	if errorMessage := stringFromAny(payload["error"], ""); errorMessage != "" {
		return errorMessage
	}
	return "no error message returned"
}

func boolFrom(value any) bool {
	boolean, ok := value.(bool)
	return ok && boolean
}

func stringFromAny(value any, fallback string) string {
	text, ok := value.(string)
	if !ok || text == "" {
		return fallback
	}
	return text
}

func mapFromAny(value any) map[string]any {
	object, ok := value.(map[string]any)
	if !ok {
		return nil
	}
	return object
}

File: examples/middleware/laravel/src/Support/AportClient.php
<?php

namespace Aport\Laravel\Support;

use Aport\Laravel\Contracts\AportClient as AportClientContract;
use Aport\Laravel\DTO\VerificationResult;
use Aport\Laravel\Exceptions\AportException;
use Aport\Laravel\Exceptions\VerificationRequestException;
use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\Exception\RequestException;

class AportClient implements AportClientContract
{
    private ClientInterface $http;

    /**
     * @param array<string, mixed> $config
     */
    public function __construct(private readonly array $config = [], ?ClientInterface $http = null)
    {
        $this->http = $http ?: new Client([
            'base_uri' => rtrim((string) ($this->config['base_url'] ?? 'https://aport.io'), '/'),
            'timeout' => (float) ($this->config['timeout'] ?? 5.0),
        ]);
    }

    /**
     * @param array<string, mixed> $context
     */
    public function verify(string $policy, string $agentId, array $context = []): VerificationResult
    {
        $policy = trim($policy);
        $agentId = trim($agentId);

        if ($policy === '') {
            throw new AportException('APort policy is required.');
        }

        if ($agentId === '') {
            throw new AportException('APort agent ID is required.');
        }

        try {
            $response = $this->http->request('POST', "/api/verify/policy/{$policy}", [
                'headers' => $this->headers(),
                'json' => [
                    'policy_id' => $policy,
                    'agent_id' => $agentId,
                    'context' => $context,
                ],
            ]);
        } catch (RequestException $exception) {
            throw $this->requestException($exception);
        } catch (GuzzleException $exception) {
            throw new VerificationRequestException(
                'Unable to reach APort verification service: ' . $exception->getMessage(),
                previous: $exception,
            );
        }

        $payload = json_decode((string) $response->getBody(), true);

        if (! is_array($payload)) {
            throw new VerificationRequestException('APort returned an invalid JSON response.');
        }

        return VerificationResult::fromArray($payload, $policy, $agentId);
    }

    /**
     * @return array<string, string>
     */
    private function headers(): array
    {
        $headers = [
            'Accept' => 'application/json',
            'Content-Type' => 'application/json',
            'User-Agent' => 'aporthq-laravel-aport/1.0',
        ];

        $apiKey = $this->config['api_key'] ?? null;

        if (is_string($apiKey) && $apiKey !== '') {
            $headers['Authorization'] = 'Bearer ' . $apiKey;
        }

        return $headers;
    }

    private function requestException(RequestException $exception): VerificationRequestException
    {
        $response = $exception->getResponse();
        $message = 'APort verification request failed.';

        if ($response !== null) {
            $body = json_decode((string) $response->getBody(), true);

            if (is_array($body) && isset($body['message'])) {
                $message = (string) $body['message'];
            } else {
                $message = sprintf(
                    'APort verification request failed with HTTP %s.',
                    $response->getStatusCode(),
                );
            }
        }

        return new VerificationRequestException($message, previous: $exception);
    }
}
